pkg/promtail/promtail.go

```diff
@@ -35,33 +34,44 @@ func New(cfg config.Config) (*Promtail, error) {
 		return nil, err
 	}
 
-	tms, err := targets.NewTargetManagers(util.Logger, positions, client, cfg.ScrapeConfig, &cfg.TargetConfig)
+	promtail := &Promtail{
+		client: client,
+	}
+
+	tms, err := targets.NewTargetManagers(promtail, util.Logger, cfg.PositionsConfig, client, cfg.ScrapeConfig, &cfg.TargetConfig)
 	if err != nil {
 		return nil, err
 	}
-
+	promtail.targetManagers = tms
 	server, err := server.New(cfg.ServerConfig, tms)
 	if err != nil {
 		return nil, err
 	}
-
-	return &Promtail{
-		client:         client,
-		positions:      positions,
-		targetManagers: tms,
-		server:         server,
-	}, nil
+	promtail.server = server
+	return promtail, nil
 }
 
 // Run the promtail; will block until a signal is received.
 func (p *Promtail) Run() error {
+	p.mtx.Lock()
+	// if we stopped promtail before the server even started we can return without starting.
+	if p.stopped {
+		return nil
```

Comment:
Is it a problem that this would cause the mutex to never unlock?

Reply:
yes we should also unlock in the if branch, good catch. However we can't use a defer cause p.server.Run() is a blocking call.